### PR TITLE
allow use of different Key Share for ESNI and handshake

### LINKED ISSUE #67: Use different client key shares for ESNI and TLS KE?
Using the same key share looked like a good idea, but it's actually a real pain to implement because you need to get pretty involved with the TLS 1.3 key exchange.

Another approach would be to have separate keys but bind the ClientHello.KeyShare extension into the ESNI extension. I had thought of putting a hash in the extension plaintext but @martinthomson points out that if you use it as AAD, then the client-facing server will automatically detect mismatch

### PR DESCRIPTION
**Pull request by kazuho**
As pointed out in #69, current draft allows a client-facing server to downgrade the TLS handshake under Split mode.

For example, if the hidden server supports X25519 and X448 but the ESNIKeys published by the client-facing server only includes X25519, client only sends X25519.

This PR fixes the issue in the least intrusive way, by allowing the parties to use different key shares for ESNI and TLS key schedule.

In short, KE for ESNI is selected by the server offering the list of supported groups in ESNIKeys and client choosing one; KE for key schedule is selected exactly the same way as is defined in TLS 1.3.

fixes #67

### CODE REVIEW

**Comment by ekr** [comment]
@kazuho: I haven't gone through this in detail, but doesn't this create the possibility of a cut-and-paste attack if the client uses a less-preferred group.

Consider the case where the server's preferences are [X448, X25519] and the client's preferences are [X25519, X448]. The client sends one key share: X25519. The attacker then adds his own X448 share. The server decrypts SNI but sends an X448 share, thus allowing the attacker to decrypt the certificate. Is there a reason why this won't work?

**Comment by kazuho** [comment]
@ekr I think you are correct. Thank you for pointing that out. I now understand that the approach used in this PR does not work.

I also notice that under Split mode, KEs and cipher-suites that can be offered in ESNIKeys will be the intersection of those supported by the various hidden servers. That would mean that the operator of the client-facing server will need to instruct the operators of hidden servers which KE and cipher-suites that they must support; otherwise only secp256r1 / aes128gcm-sha256 will be usable in practice.
